common/types/string_test.go

```diff
@@ -253,3 +253,101 @@ func TestStringSize(t *testing.T) {
 		t.Error("String size must be code points, not UTF8 bytes")
 	}
 }
+
+func TestStringFormat(t *testing.T) {
+	tests := []struct {
+		name           string
+		format         string
+		formatArgs     []ref.Val
+		err            string
+		expectedOutput string
+	}{
+		{
+			name:           "no-op",
+			format:         "no substitution",
+			expectedOutput: "no substitution",
+		},
+		{
+			name:           "int substitution",
+			format:         "int is {0}",
+			formatArgs:     []ref.Val{Int(1)},
+			expectedOutput: "int is 1",
+		},
+		{
+			name:           "double substitution",
+			format:         "double is {0}",
+			formatArgs:     []ref.Val{Double(0.5)},
+			expectedOutput: "double is 0.5",
+		},
+		{
+			name:           "mid-string substitution",
+			format:         "str is {0} and some more",
+			formatArgs:     []ref.Val{String("filler")},
+			expectedOutput: "str is filler and some more",
+		},
+		{
+			name:           "bracket escaping",
+			format:         "{{ and also }}",
+			expectedOutput: "{ and also }",
+		},
+		{
+			name:           "multiple substitutions",
+			format:         "{0} {1} {2}, {3} {4} {5}",
+			formatArgs:     []ref.Val{Int(1), Int(2), Int(3), String("A"), String("B"), String("C")},
+			expectedOutput: "1 2 3, A B C",
+		},
+		{
+			name:   "unmatched left bracket",
+			format: "filler {",
+			err:    "error while parsing arg index: could not find matching '}'",
+		},
+		{
+			name:   "unmatched right bracket",
+			format: " } filler",
+			err:    "unmatched '}'",
```

Comment:
Handle and test `unmatched '{'` case as well?

Reply:
Right, fixed